core/src/main/java/org/apache/iceberg/util/PartitionMap.java

```diff
@@ -0,0 +1,265 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.util;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+
+/**
+ * A map that uses a pair of spec ID and partition tuple as keys.
+ *
+ * <p>This implementation internally stores provided partition tuples in {@link StructLikeMap} for
+ * consistent hashing and equals behavior. This ensures that objects of different types that
+ * represent the same structs are treated as equal keys in the map.
+ *
+ * <p>Note: This map is not designed for concurrent modification by multiple threads. However, it
+ * supports safe concurrent reads, assuming there are no concurrent writes.
+ *
+ * <p>Note: This map does not support null pairs but supports null as partition tuples.
+ *
+ * @param <V> the type of values
+ */
+public class PartitionMap<V> implements Map<Pair<Integer, StructLike>, V> {
+
+  private final Map<Integer, PartitionSpec> specs;
+  private final Map<Integer, Map<StructLike, V>> partitionMaps;
+
+  private PartitionMap(Map<Integer, PartitionSpec> specs) {
+    this.specs = specs;
+    this.partitionMaps = Maps.newHashMap();
+  }
+
+  public static <T> PartitionMap<T> create(Map<Integer, PartitionSpec> specs) {
+    return new PartitionMap<>(specs);
+  }
+
+  @Override
+  public int size() {
+    return partitionMaps.values().stream().mapToInt(Map::size).sum();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return partitionMaps.values().stream().allMatch(Map::isEmpty);
+  }
+
+  @Override
+  public boolean containsKey(Object key) {
+    if (key instanceof Pair) {
+      Object first = ((Pair<?, ?>) key).first();
+      Object second = ((Pair<?, ?>) key).second();
+      if (first instanceof Integer && second instanceof StructLike) {
+        return containsKey((Integer) first, (StructLike) second);
+      }
+    }
+
+    return false;
+  }
+
+  public boolean containsKey(int specId, StructLike struct) {
+    Map<StructLike, V> partitionMap = partitionMaps.get(specId);
+    return partitionMap != null && partitionMap.containsKey(struct);
+  }
+
+  @Override
+  public boolean containsValue(Object value) {
+    return partitionMaps.values().stream().anyMatch(map -> map.containsValue(value));
+  }
+
+  @Override
+  public V get(Object key) {
+    if (key instanceof Pair) {
```

Comment:
Optional we can throw a NPE on "null" key, API doesn't say we have to but this map doesn't allow for a null key so we could do it

Reply:
Let me check built-in maps that don't support null keys. We will throw an NPE in `put`. I am not sure what would be the best behavior on get so I followed `PartitionSet`.

Reply:
Some built-in maps actually throw an exception but mainly because they call `hashCode`. I don't mind adding an extra check to see if the value is null.

Do you think that help indicate something went off, @RussellSpitzer?

Reply:
We will probably need to do the same for `remove`.

Reply:
I was just following the Java API description which says to throw NPE's if you don't support null keys but I really don't care either way. I think in our case we have no use case for a null pair so we probably should just not allow it?

Reply:
Yeah, I'll add an exception to `get` and `remove`.